Add prometheus alerting consistency guide (#10)
operate-first/SRE: prometheus_alerting_consistency.md

# Prometheus Alerting Consistency

## Summary

This guide is originally based on the [Alerting Consistency][1] proposal for OpenShift alerts.

> Clear and actionable alerts are a key component of a smooth operational experience.

Reviewer: I think there is a need to define the term `actionable` here as it can be interpreted differently by different teams/users. Such as does actionable means it has a SOP, or what are the basic requirements to determine that an alert is actionable?

Reviewer 2: Agreed. The general principle is that there should be something for an SRE to "do" to resolve the issue. That action may be following a SOP, but the SOP must contain actions that are achievable and specific. It is not sufficient to say: "logs at the logs and determine the next action". It should say something like: "look at the logs and if it contains this error then run this command". If an SRE can do something then it shouldn't be an alert.

Author: ```suggestion
Clear and actionable alerts are a key component of a smooth operational experience. If an alert doesn't have a proven documented procedure to follow when it fires, it can lead to an increase in mean time to repair (MTTR). 
```
@pepedocs @brizrobbo Does this sound reasonable? Feel free to suggest alternate wording.
I think what you 're seeing is the ideal scenario and what should be strived for. 
In reality, I think we all know that's it's less clear cut and there's going to be an element of troubleshooting with some alerts (otherwise why would you have an alert with a clear fix documented rather than get the problem fixed or automate it if it's toil).

Reviewer 2: Great @david-martin .... I'm nit-picking, but what about adding the word "proven" in there  like this: ...a **proven** documented procedure..."? Not necessary, but I was just trying to capture the idea that the process must be tested.

Reviewer: @david-martin :+1:

Ensuring you have clear and concise guidelines for engineers and SRE creating new alerts will result in a better experience for end users.

## Recommended Reading

A list of references on good alerting practices:

* [Google SRE Book - Monitoring Distributed Systems][2]
* [Prometheus Alerting Documentation][3]
* [Alerting for Distributed Systems][4]

## Style Guide

> 

Reviewer 2: Maybe (a very loose maybe :) a worthy addition as a tip: we have also found it useful for us to name the alert with the same name as the SOP. That is as a 1-1 relationship. This helps us find the SOP if the there is some "problem" with the sop_url field. Or in the case where we see the alert firing in alertmanager but it has not yet sent its notification.

Author: @brizrobbo Do you have an example?
Which came first in this case, the SOP or the Alert? (Was the SOP named after the alert?)
The reason I'm asking, if I understand what you're saying correctly, is because it would introduce a different naming convention that highlights the 'fix' rather than the 'problem'

Reviewer 2: For examples, all the filenames for the RHOAM alerts follow this convention. 
In all cases the alert was named first, and then the SOP simply mirrored that.
There is one complication: what if n-alerts followed the same SOP? For us, that is OK, as we create a SOP per alert anyway even if it is the same .... this makes it easy to differentiate it later if we need to....however, it does create an extra maintenance overhead.

Author: @brizrobbo Is this something you want to include in the consistency guide and push for other teams to use?

Reviewer 3: Recommend we leave this detail out for now.  This feels like something we'd implement if alerts, SOPs etc were auto-generated by some tooling.

* Alert names MUST be CamelCase, e.g.: `BackendServiceStuck`
* Alert names SHOULD be prefixed with a component, e.g.: `ZookeeperPersistentVolumeFillingUp`
  * There may be exceptions for some broadly scoped alerts, e.g.: `TargetDown`
* Alerts MUST include a `severity` label indicating the alert's urgency.
  * Valid severities are: `critical` or `warning` — see below for
    guidelines on writing alerts of each severity.
* Alerts MUST include `summary` and `description` annotations.
  * Think of `summary` as the first line of a commit message, or an email
    subject line.  It should be brief but informative.  The `description` is the
    longer, more detailed explanation of the alert.
* Alerts SHOULD include a `namespace` label indicating the source of the alert.
  * Many alerts will include this by virtue of the fact that their PromQL
    expressions result in a namespace label.  Others may require a static

>     namespace label.

Reviewer 4: If originating from a Kubernetes cluster, alerts SHOULD include a `cluster` label indicating the source of the alert.

Author: Oh, that's interesting. 
I'm initially thinking it's only relevant if doing federation from multiple clusters with a label for the cluster id.
However, it's also useful if you have separate alerts per cluster, but they all eventually page the same SRE team.
The label name could be different for everyone, but we can capture the intention.
How about this?
```suggestion
* Alerts SHOULD include a `namespace` label indicating the source of the alert.
  * Many alerts will include this by virtue of the fact that their PromQL
    expressions result in a namespace label.  Others may require a static
    namespace label.
* Alerts SHOULD include a cluster identifier label, especially in cases where alerts are defined or routed centrally for multiple services & clusters.
  * This can hint to an SRE where the affected service is located. It can also be useful for aggregation & filtering when analysing groups of firing alerts.
```

> * All alerts MUST include a `sop_url` annotation.

Reviewer 2: ... which directs the SRE to the location of the SOP (or runbook) that is specific to fixing that problem.

Author: ```suggestion
* All alerts MUST include a `sop_url` annotation which directs the SRE to the location of the SOP (or runbook) that is specific to fixing that problem.
```

Reviewer 5: @david-martin I might say "which directs the responder(s) to the location..." What do you think?

Reviewer 6: Why `sop_url`, instead of using the upstream `runbook_url` (e.g. as discussed [here][1])?  If an alert has both an `sop_url` and a `runbook_url`, how does an SRE know which to check first, and if/when to check the other?

[1]: https://github.com/prometheus-operator/kube-prometheus/issues/731

Author: @wking I am impartial here. I've mainly seen `sop_url`, but have also seen `runbook_url`.
Is the defacto out in the wild `runbook_url`?
We can easily change it, and add an explanation saying it's the existance of a field that links to what to do that is more important than the name of the field.

Author: Alternate wording suggestion:
```suggestion
* All alerts MUST include an annotation (e.g. `runbook_url`) which directs the responder(s) to the location of the runbook (or SOP) that is specific to fixing that problem.
```

  * Standard Operating Procedure (SOP) style documentation for resolving alerts is required. (SOP Template Placeholder - https://issues.redhat.com/browse/SIGSRE-55)

## Critical Alerts

Critical alerts are for alerting current and impending disaster situations.  These alerts page an SRE.  The situation should warrant waking someone in the middle of the
night.

> Timeline: ~5 minutes.

Reviewer 2: Timeline for action:

Author: ```suggestion
Timeline for action: ~5 minutes.
```

Reviewer 5: What does "timeline for action" mean?

Reviewer 7: does this mean needs to be responded to/acknowledged within 5 minutes of firing?

Author: Timeline removed in favour of text about severity only.

Reserve critical level alerts only for reporting conditions that may lead to service unavailability.
Failures of individual components should not trigger critical level alerts in general, unless they would result in that condition.
Configure critical level alerts so they fire before the situation becomes irrecoverable.

Example critical alert:

```yaml
- alert: BackendServiceStuck
  expr: backend_state != 1
  for: 10m
  labels:
    severity: critical
  annotations:
    summary: 'The backend service is stuck in a non-ready state'

>     description: 'The backend service {{ $labels.name }} in the {{ $labels.resource_namespace }} namespace, managed by operator {{ $labels.pod }} has been in a non-ready state for 10 minutes'

Reviewer 4: resource_namespace or namespace?

Author: I'll switch to namespace. Likely some relabeling happening in the example I copied originally

Author: ```suggestion
    description: 'The backend service {{ $labels.name }} in the {{ $labels.namespace }} namespace, managed by operator {{ $labels.pod }} has been in a non-ready state for 10 minutes'
```

    sop_url: 'https://example.com/backend_service_stuck.asciidoc'
```

This alert fires if a backend service has *not* been ready for the last 10 minutes.
This is a clear example of a critical issue that represents a threat to the operability of the service, and likely warrants paging someone.
The alert has a clear summary and description annotations, and it links to a SOP with information on investigating and resolving the issue.

The group of critical alerts should be small, very well-defined, highly documented, polished and with a high bar set for entry.

> 

Reviewer 2: Tip: if it affects your SLO, then it should be a critical alert.

Author: Ooh, interesting point.
Is this always the case? 

Let's say the timeline for action is 5 minutes for a critical alert, and we have an SLO alert firing that says the error budget will be exhausted in a day (at the current rate). We're using multi-window multi-burn rate alerts as detailed in the google sre workbook.
Should that be critical, or warning?

Reviewer 2: That is a good point yourself :) - maybe leave it out... as the "criticalness" of an alert needs to be carefully considered for each and every case, and I think that most would naturally draw the conclusion that if "t affects your SLO, then it should be a critical alert" but also leaves you free to strategically set to warning in the case that the urgency is not there.

## Warning Alerts

The vast majority of alerts should use this severity.
Issues at the warning level should be addressed in a timely manner, but don't pose an immediate threat to the operation of the service as a whole.

> Timeline: ~60 minutes

Reviewer 2: Timeline for action:

Author: ```suggestion
Timeline for action: ~60 minutes
```

Reviewer 5: Similar to above: To what is the "timeline for action" referring?

Author: My loose interpretation is that if you don't do something within this timeline, you'll likely have service unavailability.
I guess it's meant to capture the immediacy of the alert.
In these examples it's tying a timeline to a severity. 
Is that a reasonable thing to do?

Reviewer 5: Is immediacy different than severity? Would it be different for each alert? (If so, wouldn't it be better put into the runbook/sop?)

Reviewer 7: i think ~60m is too optimistic - I would say ~24 hours. Ideally warnings do not page and are triaged and trend analysis - if they become criticals consistently, that trend should be recognized.

Reviewer 8: Do we have to mention a timeline for the reaction? I don't believe we should prescribe or mention any number, because these numbers usually vary from team to team.

Author: So right now the doc is tying the severity with a specific response timeline.

* Warning => 60mins
* Critical => 5mins

I'm thinking it would be sufficient to leave it ambiguous on the exact timeline (because of reasons discussed above) and just leave this existing wording:

* Warning => Issues at the warning level should be addressed in a timely manner, but don't pose an immediate threat to the operation of the service as a whole.
* Critical => Critical alerts are for alerting current and impending disaster situations...

@dastergon @lisa If you're not objectionable I'll make this change

Reviewer 8: sgtm @david-martin

> If your alert does not meet the criteria in "Critical Alerts" above, it belongs to the warning level or lower.

Reviewer 4: "or lower" - according to the Style guide, there are only two valid severities - critical and warning

Author: Good catch.
Are there use cases for alerting severity outside of critical & warning?
I can't think of any in the scope of running a service.
Maybe some teams use other severity or lack of severity to mean something else e.g. these alerts are ignored and for informational purposes only or trialing a new alert.
I think that would be outside the scope of this guide though.

Reviewer 4: I also can't think of any. We only ever used those two (critical/warning), and we didn't follow any documentation 😂.

Reviewer 2: We have a third level, called "Info". There is no notification added to this one. It's primary purpose is to help us "understand" if something is happening, but that it is not necessarily important for us to know. An example is if a service has hit some sort of quota threshold which might mean it is being throttled. As SREs, we don't necessarily care if a service is being throttled or not, but Support does because they may get a call from a customer (in the case where the customer doesn't understand also).

Author: @brizrobbo  How does this sound if we leave out the last part of the sentence?
```suggestion
If your alert does not meet the criteria in "Critical Alerts" above, it belongs to the warning level.
```
Since we're trying to be prescriptive, I think this is reasonable from an SRE point of view.
If there is other information that 'could' be represented in alerts, like quota throttling, I'm wondering if that's a symptom of that information not being surfaced elsewhere for that audience e.g. Somewhere in the UI for the customer, or a dashboard for your support team.

Reviewer 2: That's fine @david-martin ... I don't think we need to overcomplicate it. And you 100% correct about "other information" that probably shouldn't be in alerts and can be surfaced elsewhere....however, in our experience, there have been cases where knowing XYZ has helped us debug a certain issue, and having those alerts "together" has been very helpful. That being said, I don't think it is necessary to mention that here as it does complicate it.

  * There may be exceptions for some broadly scoped alerts, e.g.: `TargetDown`
* Alerts MUST include a `severity` label indicating the alert's urgency.
  * Valid severities are: `critical` or `warning` — see below for
    guidelines on writing alerts of each severity.
* Alerts MUST include `summary` and `description` annotations.
  * Think of `summary` as the first line of a commit message, or an email
    subject line.  It should be brief but informative.  The `description` is the
    longer, more detailed explanation of the alert.
* Where applicable, alerts SHOULD include a `namespace` label indicating the source of the alert.
  * Many alerts will include this by virtue of the fact that their PromQL
    expressions result in a namespace label.  Others may require a static
    namespace label.
* Alerts SHOULD include a cluster identifier label, especially in cases where alerts are defined or routed centrally for multiple services & clusters.
  * This can hint to an SRE where the affected service is located. It can also be useful for aggregation & filtering when analysing groups of firing alerts.

> * All alerts SHOULD include an annotation (e.g. `runbook_url`) which directs the responder(s) to the location of the runbook (sometimes called a Standard Operating Procedure or SOP) that is specific to fixing that problem.

Reviewer 3: Is this pattern in wide spread use already?

Reviewer 3: As I mentioned later in this review, when there's an SLO related to the alert, we probably should mention the SLO in the alert description.

Author: > Is this pattern in wide spread use already?

It's based on the guide for Managed Openshift, so I would expect it to be in use there.
Managed Kafka has a version that is very similar (inspired by Managed Openshift).
SRES would use the `runbook_url` for sure, though it is `sop_url` in at least 1 group.
This would cover a lot of teams & services involved with prometheus based alerts.

```

This alert fires if a backend service has *not* been ready for the last 10 minutes.
This is a clear example of a critical issue that represents a threat to the operability of the service, and likely warrants paging someone.
The alert has a clear summary and description annotations, and it links to a runbook with information on investigating and resolving the issue.

The group of critical alerts should be small, very well-defined, highly documented, polished and with a high bar set for entry.

## Warning Alerts

The vast majority of alerts should use this severity.
Issues at the warning level should be addressed in a timely manner, but don't pose an immediate threat to the operation of the service as a whole.
If your alert does not meet the criteria in "Critical Alerts" above, it belongs to the warning level.

Use warning level alerts for reporting conditions that may lead to inability to deliver individual features of the service, but not the service as a whole.
Most alerts are likely to be warnings.
Configure warning level alerts so that they do not fire until components have sufficient time to try to recover from the interruption automatically.

> You should have a policy for triaging warning alerts and responding in a timely manner e.g. issues created automatically on a backlog that is being actively worked from.

Reviewer 3: The text we end up with here should try to match what teams already do (or plan to do asap).  Here's another draft ADR:  https://docs.google.com/document/d/1j2F8HgqYGsTnoq1fii2ZMGacuC6jYUp0wsDOVDXL1Tc @dofinn

Author: Agreed.
I've requested access to that doc to see how well these align.
I'll defer to @dofinn if there's something to align better in the mean time.

Reviewer 7: this doc is incomplete.

Author: @dofinn Is there anything you'd like to add here at this time? 
I see this consistency guide complimenting the ADR you're working on.
If you're planning on adding the ADR to this repo, we could link the docs

Reviewer 7: this looks pretty good to me :) i find this such a subjective issue, but this is a good happy medium

Example warning alert:

```yaml
- alert: PrometheusDuplicateTimestamps
  annotations:

>     description: Prometheus {{$labels.namespace}}/{{$labels.pod}} is dropping

Reviewer 3: Same comment as for critical alerts. If there's an SLO related to the alert, mention it in the description.

# Prometheus Alerting Consistency

## Summary

This guide is originally based on the [Alerting Consistency][1] proposal for OpenShift alerts.

Clear and actionable alerts are a key component of a smooth operational experience.
Ensuring you have clear and concise guidelines for engineers and SRE creating new alerts will result in a better experience for end users.

## Recommended Reading

A list of references on good alerting practices:

* [Google SRE Book - Monitoring Distributed Systems][2]
* [Prometheus Alerting Documentation][3]
* [Alerting for Distributed Systems][4]

## Style Guide

> * Alert names MUST be CamelCase, e.g.: `BackendServiceStuck`

Reviewer 5: Nit: If you're going to use language from RFC 2119 you should call that out explicitly, and provide a link to the RFC.

> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL
> NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and
> "OPTIONAL" in this document are to be interpreted as described in
> [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).

Author: That's a good idea. I haven't seen or noticed it in any other documents before, but it makes sense to clarify wording.

> * Alert names SHOULD be prefixed with a component, e.g.: `ZookeeperPersistentVolumeFillingUp`

Author: ```suggestion
* Alert names SHOULD be prefixed with a component, e.g.: `ZookeeperPersistentVolumeFillingUp`, or an SLO name if the alert is a [Multi-Window, Multi-Burn Rate alert](https://sre.google/workbook/alerting-on-slos/#6-multiwindow-multi-burn-rate-alerts)
```

  * There may be exceptions for some broadly scoped alerts, e.g.: `TargetDown`
* Alerts MUST include a `severity` label indicating the alert's urgency.
  * Valid severities are: `critical` or `warning` — see below for
    guidelines on writing alerts of each severity.
* Alerts MUST include `summary` and `description` annotations.
  * Think of `summary` as the first line of a commit message, or an email
    subject line.  It should be brief but informative.  The `description` is the
    longer, more detailed explanation of the alert.

> * Alerts SHOULD include a `namespace` label indicating the source of the alert.

Reviewer 5: ```suggestion
* Where applicable, alerts SHOULD include a `namespace` label indicating the source of the alert.
```

SLOs have value outside clusters. :)

  * Many alerts will include this by virtue of the fact that their PromQL
    expressions result in a namespace label.  Others may require a static
    namespace label.
* Alerts SHOULD include a cluster identifier label, especially in cases where alerts are defined or routed centrally for multiple services & clusters.
  * This can hint to an SRE where the affected service is located. It can also be useful for aggregation & filtering when analysing groups of firing alerts.

> * All alerts MUST include a `sop_url` annotation.

Reviewer 5: I suggest making this "SHOULD" and maybe calling it something more generic than `sop_url`. `runbook_url`, `resolution_tips_link`. If none other those make sense, I'm 👍 to https://github.com/operate-first/sre/pull/10/files#r849645118 (though with my note on that thread) to explain what a `sop_url` is for; what is "SOP?" .

  * Standard Operating Procedure (SOP) style documentation for resolving alerts is required. (SOP Template Placeholder - https://issues.redhat.com/browse/SIGSRE-55)

## Critical Alerts

Critical alerts are for alerting current and impending disaster situations.  These alerts page an SRE.  The situation should warrant waking someone in the middle of the
night.

Timeline: ~5 minutes.

Reserve critical level alerts only for reporting conditions that may lead to service unavailability.

> Failures of individual components should not trigger critical level alerts in general, unless they would result in that condition.

Reviewer 6: Components may be composed of several services.  Probably worth describing what "service unavailability" means, if it's expected to be bigger than individual components.  Is it just Kube API service uptime?  Something else?

Author: Interesting thought.
Isn't service unavailability different for each service though?
It's an important thing that you define and measure as SLIs, so they are inherently different for every service (though patterns do exist).

Reviewer 7: most subjective comments in this space should be solved by answering "If you were a customer of the service, is there an current or immenent condition that degrades your experience with the service?" this traverses scopes...

Author: @wking any more thoughts on the above?

Reviewer 3: The SRE-P team has a draft ADR for this, SD-ADR-0052 internally to agree upon this pattern "officially". @dofinn

Reviewer 9: Another perspective: I think individual components' unavailability should still trigger paging alerts if they're in the user happy path and thus a part of the SLO.

Author: @aditya-konarde so, replace with this?
```suggestion
Failures of individual components should trigger paging alerts if they're in the user happy path and thus a part of the SLO.
```

Reviewer 10: I’m taking my first look through here, and I’d say “user happy path” is definitely jargonny. How about “Failures of individual components should only trigger paging alerts if they are likely to affect a related SLO”?

Configure critical level alerts so they fire before the situation becomes irrecoverable.

Example critical alert:

```yaml
- alert: BackendServiceStuck
  expr: backend_state != 1
  for: 10m
  labels:
    severity: critical
  annotations:
    summary: 'The backend service is stuck in a non-ready state'

>     description: 'The backend service {{ $labels.name }} in the {{ $labels.namespace }} namespace, managed by operator {{ $labels.pod }} has been in a non-ready state for 10 minutes'

Reviewer 6: This explains what's going on, but doesn't make it immediately obvious why the SRE is expected to care, or what the SRE is expected to do about it.  Maybe we can assume that "what next?" is covered in the SOP, but I expect the alert description to at least hint at why folks should care, to help them rank several competing alerts when deciding which SOP to read first.

Author: @wking What would you suggest as a good description in this example? (Doesn't have to be real, just capture what you're trying to highlight)
If I'm understanding you correctly, my own suggestion would be something like
```
The backend service {{ $labels.name }} in the {{ $labels.namespace }} namespace, managed by operator {{ $labels.pod }} has been in a non-ready state for 10 minutes. If this service stays in this state, API calls will fail and users won't be able to manage their resources. Managing resources is a core feature of the service.
```

Reviewer 7: as an idea (for criticals only) - should we mandate the word "customer" in descriptions so their perspective/experience HAS to be suggested/predicted in some form.

Author: @wking any more thoughts on the above?

Reviewer 3: In the definition of each alert, should we put the SLO that it supports (if there is one) to the description?  Then an SRE would immediately know "why".

The group of critical alerts should be small, very well-defined, highly documented, polished and with a high bar set for entry.

## Warning Alerts

The vast majority of alerts should use this severity.
Issues at the warning level should be addressed in a timely manner, but don't pose an immediate threat to the operation of the service as a whole.

Timeline: ~60 minutes

If your alert does not meet the criteria in "Critical Alerts" above, it belongs to the warning level or lower.

Use warning level alerts for reporting conditions that may lead to inability to deliver individual features of the service, but not the service as a whole.
Most alerts are likely to be warnings.
Configure warning level alerts so that they do not fire until components have sufficient time to try to recover from the interruption automatically.
You should have a policy for triaging warning alerts and responding in a timely manner e.g. issues created automatically on a backlog that is being actively worked from.

Example warning alert:

```yaml
- alert: ZookeeperPersistentVolumeFillingUp

>   expr: (kubelet_volume_stats_available_bytes{persistentvolumeclaim=~"data-(.+)-zookeeper-[0-9]+"} / kubelet_volume_stats_capacity_bytes{persistentvolumeclaim=~"data-(.+)-zookeeper-[0-9]+"} < 0.15) and predict_linear(kubelet_volume_stats_available_bytes{persistentvolumeclaim=~"data-(.+)-zookeeper-[0-9]+"}[6h], 4 * 24 * 3600) < 0

Reviewer 5: I might pick something with a less complex PromQL expression, this one is enough to make someone go "wait what?" 😉 (The focus should be on the structure of the YAML, not so much on the expression inside of it.)

Author: Agreed. 
I'll change this to a simpler expression that shouldn't shift focus away from what's important in this doc i.e. consistency

  for: 1h
  labels:
    severity: warning
  annotations:
    summary: 'Zookeeper PersistentVolume is filling up.'

>     description: 'Based on recent sampling, the Zookeeper PersistentVolume claimed by {{ $labels.persistentvolumeclaim }} in Namespace {{ $labels.namespace }} is expected to fill up within four days. Currently {{ $value | humanizePercentage }} is available.'

Reviewer 6: Another "why do I care?" nit: what happens if the volume fills up?  If I have a bunch of fires going on, how do I rank the risk of this against those?

Author: Great question.
I can't think how you can answer it without getting into the specifics of a service.
Although the ability to inhibit a rule exists, so you could structure you rules so that alerts that you know are the cause of another set of alerts firing can be prioritised.
Is that worth mentioning here?

Author: Any further thoughts on this @wking ?

Reviewer 10: “Storage full” is one of those bread and butter alerts which should be standardised as much as possible. I’m not such a great fan of “disk is filling” alerts as they’re hard to tune and prone to making unnecessary noise, but if anyone wants to use them (with care) the above seems fine. But it’s just another tool in the box, not necessary an essential component.

The other thing which makes me leery of “disk filling” is that during any period of abnormal operation disk usage patterns inevitably change (logs can’t be flushed from temporary storage, queued messages stack up in the spool, all that stuff) and that’s totally normal - so alerting on it as a routine matter when other stuff is on fire is likely to be a red herring.

“Disk full” (or >98% or whatevs) *is* a useful alert because it’s solidly actionable and a good starting point for both debugging and long-term remediation of an issue, gives hints on future growth and capacity planning, etc etc.

    sop_url: 'https://example.com/persistent_volume_filling.asciidoc'
```

This alert fires if one or more Zookeeper volumes are getting close to filling up.
The alert has a clear name and informative summary and description annotations, and it links to a SOP with information on investigating and resolving the issue.

> The timeline is appropriate for allowing the service to resolve the issue itself, avoiding the need to alert SRE.

Reviewer 6: hmm, `for: 1h` seems like a good way to keep from bothering humans for short-term effects.  But I'm having a bit of trouble wrapping my head around how it interacts with the 6h range being used for [`predict_linear`][1].  Might be worth checking with the monitoring folks to see if there are recommendations like "when using `predict_linear`, you don't need to bother with `for`" or "when using `predict_linear`, set `for` to something larger than the range given to `predict_linear`.

[1]: https://prometheus.io/docs/prometheus/latest/querying/functions/#predict_linear

Reviewer 11: I think both time ranges (one for `for`, one for `predict_linear`) serve different purposes and work well in concert.

1. `predict_linear`: This time range tells prometheus which samples to select to run a simple linear regression. It will work if there are [at least 2 samples present](https://github.com/prometheus/prometheus/blob/fc3f1b8ded810a1263f72d37db9db4e8ce726120/promql/functions.go#L774-L788). So in this query, the `predict_linear` part says, "given the last 6 hours of data, the volume will fill up in 4 days or less".
2. The `for` clause keeps this from firing unless the query as been true for an hour. This normalizes big spikes in the samples considered.

An example where this matters is when you have a big spike in temporary data written, i.e. data that has a short life time. If the `for` clause is left out, a big enough spike will trigger the alert, no matter how many samples are passed to `predict_linear` (or the time range is so large that it'll never trigger). What this misses though is the when the data is deleted shortly after, the alert will already have fired.
With the `for` clause there is a chance to catch this sort of scenario.
I don't think we can give a general rule how these work together best, it depends on the signal we want an alert on.

For this example in particular I think the question is: Are we expecting spikes in temporary data? If not it might make sense to drop the `for` clause.

Additional thought:

- Maybe drop the first part of the query. This alert will only fire if `predict_linear` falls under zero _and_ the volume has only 15% space available. The last part is not mentioned in the alert name or description. I would separate that into a second alert to improve clarity.
- Be more explicit in the alert description. Why not something like `Based on the last 6 hours worth of data, the Zookeeper PersistentVolume claimed by {{ $labels.persistentvolumeclaim }} in Namespace {{ $labels.namespace }} is expected to fill up within four days or less.`

Reviewer 6: > If the `for` clause is left out, a big enough spike will trigger the alert, no matter how many samples are passed to `predict_linear`...
> ...
> For this example in particular I think the question is: Are we expecting spikes in temporary data? If not it might make sense to drop the `for` clause.

Fair, so possibly no general rule.

> Maybe drop the first part of the query. This alert will only fire if `predict_linear` falls under zero and the volume has only 15% space available.

You need the <15% bit to turn off the alert quickly after space is freed, because it might take a while after cleanup before the 6h linear regression pulls back up into a positive extrapolated prediction.

Some extreme cases for disk usage:

* Most false-happy prediction would be 0% free for 6h, a single scrape at 100% free, and then right back to 0% free.  It would take 3h after the return to 0% before the predicted future dropped back below 0%.
* Most false-sad prediction would be trending straight from 100% free to 10% free over 6h, and then settles in around 10% without dropping further.  As long as you didn't start dropping again, you'd be safe with 10% free, but the 4d prediction would be negative for several hours until the earlier very-free numbers started to drop out of the 6h `predict_linear`.

The `for: 1h` doesn't make much difference in either of those.  I guess you could have an oscillation with a <1h period, where the peak-free was over 15%, or the period was right around 6h and predicted zero time dropped under the 4h threshold, but only for <1h at a time.  I guess what I'm deciding is that the thresholds here need to be tied fairly tightly to guesses about likely behavior (e.g. "our garbage-collection routine kicks in at <15% free, and usually runs for <15m"), and selecting these values with a more generic understanding of the system being monitored is pretty risky.

Reviewer 11: > You need the <15% bit to turn off the alert quickly after space is freed, because it might take a while after cleanup before the 6h linear regression pulls back up into a positive extrapolated prediction.

Good point yeah. My main thought was that it would be nice to have a low space alert as well. If that is of interest, the filling-up alert could query the synthetic ALERTS metric and replace the first part with something like `ALERTS{alertname="ZookeeperPersistentVolumeAvailableLow",alertstate="firing", persistentvolumeclaim="foo"} = 1`

I agree that more insight into the IO behavior will help to write a more precise alert. Even then some weird IO pattern can probably make this noisy. That's why I think making the description as explicit as possible is good practice. The more information the respondent has, the better they can structure their response. 
"Based on the IO rates of the last 6 hours, you PVC will be full in $time" is much more actionable then "Your PVC is will be full soon", to exaggerate a bit.

Author: @jan--f @wking 
This conversation is excellent, and I think it would be great to capture what you think the outcome of it is.
As an fyi, I've since pushed a simpler promql expression for the example alert so as to keep the focus of the document on consistency of alerts.
However, I feel like there's some wording updates needed to capture what you're saying here in a general way,
as well as any tips for using predict_linear you think are relevant

Reviewer 12: > As an fyi, I've since pushed a simpler promql expression for the example alert so as to keep the focus of the document on consistency of alerts.

In this case I would remove all references to `predict_linear` and I would also change the example since alerting on a fixed threshold value is something I wouldn't recommend. https://github.com/openshift/cluster-monitoring-operator/blob/bc45b875b3aa4c4f36abf35b532e3bc9294b75b6/assets/prometheus-k8s/prometheus-rule.yaml#L114-L124 might be a good alternative (though it is Prometheus specific, it should be easy to grasp).

Author: Thanks for the suggestion @simonpasquier 
I've updated the example & text.

# Prometheus Alerting Consistency

## Summary

This guide is originally based on the [Alerting Consistency][1] proposal for OpenShift alerts.

Clear and actionable alerts are a key component of a smooth operational experience. If an alert doesn't have a proven documented procedure to follow when it fires, it can lead to an increase in mean time to repair (MTTR).
Ensuring you have clear and concise guidelines for engineers and SRE creating new alerts will result in a better experience for end users.

## Recommended Reading

A list of references on good alerting practices:

* [Google SRE Book - Monitoring Distributed Systems][2]
* [Prometheus Alerting Documentation][3]
* [Alerting for Distributed Systems][4]

> 

Reviewer 9: I also usually recommend this one: https://docs.google.com/document/d/199PqyG3UsyXlwieHaqbGiWVa8eMWi8zzAn0YfcApr8Q/edit

Author: 👍

## Style Guide

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).

* Alert names MUST be CamelCase, e.g.: `BackendServiceStuck`
* Alert names SHOULD be prefixed with a component, e.g.: `ZookeeperPersistentVolumeFillingUp`, or a Service Level Objective (SLO) name if the alert is a [Multi-Window, Multi-Burn Rate alert](https://sre.google/workbook/alerting-on-slos/#6-multiwindow-multi-burn-rate-alerts)

>   * There may be exceptions for some broadly scoped alerts, e.g.: `TargetDown`

Reviewer 9: Do we recommend 'TargetDown' alerts? Shouldn't those be covered by other SLI's anyway?

Author: Are you suggesting a different name or removing this bullet?

Reviewer 10: TargetDown and friends are useful backstops to have as an insurance policy against stuff breaking which isn't caught by more specific alerts. No alerting configuration is perfect..

* Alerts MUST include a `severity` label indicating the alert's urgency.
  * Valid severities are: `critical` or `warning` — see below for
    guidelines on writing alerts of each severity.

> * Alerts MUST include `summary` and `description` annotations.

Reviewer 9: As a best practice, shouldn't we move the `runbook_url` to a `MUST`?

Author: 👍

  * Think of `summary` as the first line of a commit message, or an email
    subject line.  It should be brief but informative.  The `description` is the
    longer, more detailed explanation of the alert.
* Where applicable, alerts SHOULD include a `namespace` label indicating the source of the alert.
  * Many alerts will include this by virtue of the fact that their PromQL
    expressions result in a namespace label.  Others may require a static
    namespace label.
* Alerts SHOULD include a cluster identifier label, especially in cases where alerts are defined or routed centrally for multiple services & clusters.
  * This can hint to an SRE where the affected service is located. It can also be useful for aggregation & filtering when analysing groups of firing alerts.
* All alerts SHOULD include an annotation (e.g. `runbook_url`) which directs the responder(s) to the location of the runbook (sometimes called a Standard Operating Procedure or SOP) that is specific to fixing that problem.

## Critical Alerts

> Critical alerts are for alerting current and impending disaster situations.  These alerts page an SRE.  The situation should warrant waking someone in the middle of the night.

Reviewer 9: ```suggestion
Critical alerts are for alerting current and impending disaster situations.  These alerts page an on-call Engineer.  The situation should warrant waking someone in the middle of the night and/or interrupting their regular workflow.
```

Reserve critical level alerts only for reporting conditions that may lead to service unavailability.
Failures of individual components should not trigger critical level alerts in general, unless they would result in that condition.
Configure critical level alerts so they fire before the situation becomes irrecoverable.

Example critical alert:

```yaml
- alert: BackendServiceStuck
  expr: max_over_time(backend_state[5m]) != 1
  for: 10m
  labels:
    severity: critical
  annotations:
    summary: 'The backend service is stuck in a non-ready state'
    description: 'The backend service {{ $labels.name }} in the {{ $labels.namespace }} namespace, managed by operator {{ $labels.pod }} has been in a non-ready state for 10 minutes'
    runbook_url: 'https://example.com/backend_service_stuck.asciidoc'
```

> This alert fires if a backend service has *not* been ready for the last 10 minutes.

Reviewer 9: Is the 10 minutes an arbitrary time here? If we already know the service is down, why wait for 10 minutes here? Would it be helpful to add a few sentences about the 'for' clause and how the alerting really works under the hood?

Author: In this example it's somewhat arbitrary.
Yeah, I think that might be helpful to add a few sentences.
If you can suggest something that would be great.
Feel free to suggest any changes to the example as well that help convey timing better.

Reviewer 10: “has not been ready for a set time. To avoid false alarms his time should be chosen to be long enough that it is unlikely to be reached during a routine restart or deployment”?

* Alert names MUST be CamelCase, e.g.: `BackendServiceStuck`
* Alert names SHOULD be prefixed with a component, e.g.: `ZookeeperPersistentVolumeFillingUp`, or an SLO name if the alert is a [Multi-Window, Multi-Burn Rate alert](https://sre.google/workbook/alerting-on-slos/#6-multiwindow-multi-burn-rate-alerts)
  * There may be exceptions for some broadly scoped alerts, e.g.: `TargetDown`
* Alerts MUST include a `severity` label indicating the alert's urgency.
  * Valid severities are: `critical` or `warning` — see below for
    guidelines on writing alerts of each severity.
* Alerts MUST include `summary` and `description` annotations.
  * Think of `summary` as the first line of a commit message, or an email
    subject line.  It should be brief but informative.  The `description` is the
    longer, more detailed explanation of the alert.
* Where applicable, alerts SHOULD include a `namespace` label indicating the source of the alert.
  * Many alerts will include this by virtue of the fact that their PromQL
    expressions result in a namespace label.  Others may require a static
    namespace label.

> * Alerts SHOULD include a cluster identifier label, especially in cases where alerts are defined or routed centrally for multiple services & clusters.

Reviewer 7: Should this be standardized `key` for this label? `_id` is used in telemeter and I have seen it sprinkled among HyperShift docs.

Reviewer 10: I’d shy away from using _-prefixed names as that carries the “internally defined label” vibe. How about just suggesting, say, `cluster`?

  * This can hint to an SRE where the affected service is located. It can also be useful for aggregation & filtering when analysing groups of firing alerts.
* All alerts SHOULD include an annotation (e.g. `runbook_url`) which directs the responder(s) to the location of the runbook (sometimes called a Standard Operating Procedure or SOP) that is specific to fixing that problem.

## Critical Alerts

Critical alerts are for alerting current and impending disaster situations.  These alerts page an SRE.  The situation should warrant waking someone in the middle of the
night.

Timeline: ~5 minutes.

Reserve critical level alerts only for reporting conditions that may lead to service unavailability.
Failures of individual components should not trigger critical level alerts in general, unless they would result in that condition.
Configure critical level alerts so they fire before the situation becomes irrecoverable.

Example critical alert:

```yaml
- alert: BackendServiceStuck

>   expr: backend_state != 1

Reviewer 12: something we haven't captured (yet) in the original document is how to write expressions that don't create false-negatives on failed scrapes. E.g. this would be a better version:

```suggestion
  expr: max_over_time(backend_state[5m]) != 1
```

  for: 10m
  labels:
    severity: critical
  annotations:
    summary: 'The backend service is stuck in a non-ready state'
    description: 'The backend service {{ $labels.name }} in the {{ $labels.namespace }} namespace, managed by operator {{ $labels.pod }} has been in a non-ready state for 10 minutes'
    runbook_url: 'https://example.com/backend_service_stuck.asciidoc'
```

This alert fires if a backend service has *not* been ready for the last 10 minutes.

> This is a clear example of a critical issue that represents a threat to the operability of the service, and likely warrants paging someone.

Reviewer 7: does it impact the customer? If iv got HA  - does that lower the severity?

Reviewer 10: This is obviously something that will vary depending on whether service is HA or not, hence not prescribing a particular timeout  here.

The alert has a clear summary and description annotations, and it links to a runbook with information on investigating and resolving the issue.

The group of critical alerts should be small, very well-defined, highly documented, polished and with a high bar set for entry.

## Warning Alerts

The vast majority of alerts should use this severity.
Issues at the warning level should be addressed in a timely manner, but don't pose an immediate threat to the operation of the service as a whole.

Timeline: ~60 minutes

> If your alert does not meet the criteria in "Critical Alerts" above, it belongs to the warning level.

Reviewer 7: how do i report contextual conditions? - what about severity of "info"

Reviewer 10: There’s a pretty reasonable train of thought that “info” and “alert” don’t really go together. Put it on a dashboard if you like, but don’t put it anywhere near anyone’s pager.

* Alert names MUST be CamelCase, e.g.: `BackendServiceStuck`
* Alert names SHOULD be prefixed with a component, e.g.: `ZookeeperPersistentVolumeFillingUp`, or a Service Level Objective (SLO) name if the alert is a [Multi-Window, Multi-Burn Rate alert](https://sre.google/workbook/alerting-on-slos/#6-multiwindow-multi-burn-rate-alerts)
  * There may be exceptions for some broadly scoped alerts, e.g.: `TargetDown`
* Alerts MUST include a `severity` label indicating the alert's urgency.
  * Valid severities are: `critical` or `warning` — see below for
    guidelines on writing alerts of each severity.
* Alerts MUST include `summary` and `description` annotations.
  * Think of `summary` as the first line of a commit message, or an email
    subject line.  It should be brief but informative.  The `description` is the
    longer, more detailed explanation of the alert.
* Where applicable, alerts SHOULD include a `namespace` label indicating the source of the alert.
  * Many alerts will include this by virtue of the fact that their PromQL
    expressions result in a namespace label.  Others may require a static
    namespace label.

> * Alerts SHOULD include a cluster identifier label, especially in cases where alerts are defined or routed centrally for multiple services & clusters.

Reviewer 10: ```suggestion
* Alerts SHOULD include a cluster identifier label, especially in cases where alerts are defined or routed centrally for multiple services & clusters. This cluster identifier label SHOULD be consistent and obvious - for example, `cluster`.
```

  * This can hint to an SRE where the affected service is located. It can also be useful for aggregation & filtering when analysing groups of firing alerts.
* All alerts MUST include an annotation (e.g. `runbook_url`) which directs the responder(s) to the location of the runbook (sometimes called a Standard Operating Procedure or SOP) that is specific to fixing that problem.

## Critical Alerts

Critical alerts are for alerting current and impending disaster situations.  These alerts page an on-call Engineer.  The situation should warrant waking someone in the middle of the night and/or interrupting their regular workflow.

Reserve critical level alerts only for reporting conditions that may lead to service unavailability.

> Failures of individual components should not trigger critical level alerts in general, unless they would result in that condition.

Reviewer 10: ```suggestion
Failures of individual components MUST NOT trigger critical level alerts unless that failure is likely to affect a related SLO if not attended to.
```

Configure critical level alerts so they fire before the situation becomes irrecoverable.

Example critical alert:

```yaml
- alert: BackendServiceStuck
  expr: max_over_time(backend_state[5m]) != 1
  for: 10m
  labels:
    severity: critical
  annotations:
    summary: 'The backend service is stuck in a non-ready state'
    description: 'The backend service {{ $labels.name }} in the {{ $labels.namespace }} namespace, managed by operator {{ $labels.pod }} has been in a non-ready state for 10 minutes'
    runbook_url: 'https://example.com/backend_service_stuck.asciidoc'
```

> This alert fires if a backend service has *not* been ready for the last 10 minutes.

Reviewer 10: ```suggestion
This alert fires if a backend service has *not* been ready for a pre-configured time. To avoid false alarms the chosen timeout MUST be long enough that it is unlikely to be reached during any routine restart or deployment. 
```
